v2/internal/reconcilers/arm/kubernetes_resource_exporter.go

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package arm
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Azure/azure-service-operator/v2/internal/set"
+	asocel "github.com/Azure/azure-service-operator/v2/internal/util/cel"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/configmaps"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/secrets"
+)
+
+type kubernetesResourceExporter interface {
+	Export(ctx context.Context) ([]client.Object, error)
+}
+
+var _ kubernetesResourceExporter = &configMapExpressionExporter{}
+
+type configMapExpressionExporter struct {
+	obj                 genruntime.ARMMetaObject
+	versionedObj        genruntime.ARMMetaObject
+	expressionEvaluator asocel.ExpressionEvaluator
+}
+
+func (c *configMapExpressionExporter) Export(ctx context.Context) ([]client.Object, error) {
+	cmExporter, ok := c.obj.(configmaps.Exporter)
+	if !ok {
+		return nil, nil
+	}
+
+	cmExpressions := cmExporter.ConfigMapDestinationExpressions()
+	if len(cmExpressions) == 0 {
+		return nil, nil
+	}
+
+	resources, err := c.parseConfigMaps(cmExpressions)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse configmap expressions for export")
+	}
+	return resources, err
+}
+
+func (c *configMapExpressionExporter) parseConfigMaps(expressions []*core.DestinationExpression) ([]client.Object, error) {
+	collector := configmaps.NewCollector(c.obj.GetNamespace())
+
+	for _, expression := range expressions {
+		value, err := c.expressionEvaluator.CompileAndRun(expression.Value, c.versionedObj, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		if value.Value != "" {
+			collector.AddValue(
+				&genruntime.ConfigMapDestination{
+					Name: expression.Name,
+					Key:  expression.Key,
+				}, value.Value)
+		} else if len(value.Values) > 0 {
```

Comment:
Could you have both `value.Value != ""` and `len(value.Values) > 0` true at the same time? 

If so, the current behaviour is to use `value.Value` and ignore `value.Values` - is this the desired behaviour? Or should this be an error?

Reply:
It's not really possible to have both, if you look at the [CompileAndRun](https://github.com/Azure/azure-service-operator/blob/main/v2/internal/util/cel/cel.go#L207) function, it will only ever return either `Value` or `Values` and which of those it does is directly related to if the users CEL expression is of result type `string` or `map[string]string`.

I think here it's OK to treat the `ExpressionResult` from `CompileAndRun` as a discriminated union here (which is what it's intended to be). If there were a bug here, it would be in the `ComileAndRun` function returning two different types of output. As implemented at least today, that's not possible.

Reply:
Fair enough.